python/GeologicalModel3D.py

```diff
@@ -0,0 +1,338 @@
+# You can redistribute it and/or modify it under the terms of the GNU Affero General Public License version 3.
+#
+
+from collections import namedtuple
+import json
+import numpy as np
+import pypotential3D as pypotential
+import topography_reader
+
+Intersection = namedtuple('Intersection', ['point', 'rank'])
+Box =  namedtuple('Box', ['xmin', 'ymin', 'zmin', 'xmax', 'ymax', 'zmax'])
+
+GradientData = namedtuple('GradientData', ['locations', 'values'])
+class PotentialData:
+    @classmethod
+    def from_dict(cls, dct):
+        # Interface stuff memo
+        # input: 'interfaces': [{'points': [{'x': 0.0, 'y': 0.0, 'z': 0.0}, {'x': 1.0, 'y': 1.0, 'z': 1.0}]}],
+        # output: [array([[0.0, 0.0, 0.0], [1.0, 1.0, 1.0]])]
+        
+        def vec_to_list(vec):
+            return [vec['x'], vec['y'], vec['z']]
+
+        # The original code seems to assume that the points are a numpy array, but the list of points as a normal list/array
+        def points_to_array(points):
+            return np.array([vec_to_list(v) for v in points], dtype=np.double)
+
+        def gradiant_data(gds):
+            locations = []
+            values = []
+            for gd in gds:
+                locations += gd['position']['x'], gd['position']['y'], gd['position']['z']
+                values += gd['value']['x'], gd['value']['y'], gd['value']['z']
+                
+            return GradientData(
+                np.array(locations, dtype=np.double),
+                np.array(values, dtype=np.double)
+            )
+        
+        return PotentialData(
+            CovarianceModel.from_dict(dct['covarianceModel']),
+            gradiant_data(dct['gradients']),
+            [points_to_array(x['points']) for x in dct['interfaces']]
+        )
+
+    def __init__(self, covariance_model, gradients, interfaces):
+        self.covariance_model = covariance_model
+        self.gradients = gradients
+        self.interfaces = interfaces
+
+class SeriesData:
+    @classmethod
+    def from_dict(cls, dct):
+        res = SeriesData(dct['name'])
+        res.formations = dct['formations']
+        res.relation = list(dct['relation'].keys())[0].lower()
+        res.influenced_by_fault = dct['influencedByFaults']
+        res.potential_data = PotentialData.from_dict(dct['potentialData'])
+        return res
+
+    def __init__(self, name):
+        self.name = name
+        self.formations = None
+        self.relation = None
+        self.influenced_by_fault = None
+        self.potential_data = None
+
+class Pile:
+    @classmethod
+    def from_dict(cls, dct):
+        res = SeriesData(list(dct['reference'].keys())[0].lower())
+        res.all_series = [SeriesData.from_dict(x) for x in dct['series']]
+        return res
+        
+    def __init__(self, reference):
+        assert reference=='top' or reference=='base'
+        self.reference = reference
+        self.all_series = None
+
+class CovarianceModel:
```

Comment:
I am starting to wonder now if we should not just hardcode this in python, I assume VK will always use the default values here?

Reply:
Yes, but if we someday decide to have non-hardcoded values, we would have to update the Python script for that (the changes in the Scala codebase would be less important than in this script)

Reply:
For values we are very likely to change this makes sense, for everything else I would just hardcode it in Python as it is simpler. As we are creating the Python script ourselves anyway, changing it later is no problem and this tends to simplify the interface. I suspect we are never going to change the CovarianceModel.